### PR TITLE
Add lm32 "lite" variant, remove mult/div from "minimal" and update compiler flags accordingly.

### PR DESCRIPTION
**Pull request by cr1901**
I have split the `lm32` `minimal` variant into 2 variants, called `lite` and `minimal`:

## Lite

* Same as previous `minimal`, except now with 2kb I-cache (16 bytes per line, 128 lines). This is suitable for running a program from SPI flash without the CPU spending most time waiting for instructions. No data cache, so all RAM references are 2 cycles.

## Minimal

* Same as previous `minimal`, except multiply and divide disabled, barrel shift and sign extend still present. This reduces resources used for debugging purposes.

---

This also serves as a watered-down version of a [PR](https://github.com/mithro/litex/pull/2) on @mithro's litex branch; changing compiler/CPU flags based on variant is now supported!

I have confirmed both variants to boot to the BIOS by testing on tinyfpga.

---

`mulsi3.c` was compiled but never linked into `libcompiler_rt.a`, so I fixed that as well. The `minimal` variant will choke when linking the BIOS without this.

### CODE REVIEW

File: litex/soc/cores/cpu/lm32/verilog/config_lite/lm32_config.v [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+
+//
+// REGISTER FILE
+//
+
+// The following option explicitly infers block RAM for the register
+// file. There is extra logic to avoid parallel writes and reads.
+// Normally, if your synthesizer is smart enough, this should not be
+// necessary because it will automatically infer block RAM for you.
+//`define CFG_EBR_POSEDGE_REGISTER_FILE
```
Comment (mithro) [inline litex/soc/cores/cpu/lm32/verilog/config_lite/lm32_config.v]: Do you know if Yosys is smart enough here?

File: litex/soc/cores/cpu/lm32/verilog/config_lite/lm32_config.v [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+
+// Explicitly infers block RAM, too. But it uses two different clocks,
+// one being shifted by 180deg, for the read and write port. Therefore,
+// no additional logic to avoid the parallel write/reads.
+//`define CFG_EBR_NEGEDGE_REGISTER_FILE
```
Comment (mithro) [inline litex/soc/cores/cpu/lm32/verilog/config_lite/lm32_config.v]: The iCE40 block ram's have independent read / write clocks, including the ability to use the negative edges?

File: litex/soc/cores/cpu/lm32/verilog/config_lite/lm32_config.v [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+
+
+//
+// MISCELLANEOUS
+//
+
+// Exceptions on wishbone bus errors
+//`define CFG_BUS_ERRORS_ENABLED
```
Comment (mithro) [inline litex/soc/cores/cpu/lm32/verilog/config_lite/lm32_config.v]: Let's make a note to investigate this at a later time?

File: litex/soc/cores/cpu/lm32/verilog/config_lite/lm32_config.v [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+
+// Enable the cycle counter
+`define CFG_CYCLE_COUNTER_ENABLED
```
Comment (mithro) [inline litex/soc/cores/cpu/lm32/verilog/config_lite/lm32_config.v]: Do we need this?

File: litex/soc/software/libcompiler_rt/Makefile [hunk 5]
```diff
@@ -15,7 +15,7 @@ all: libcompiler_rt.a
 
 libcompiler_rt.a: $(OBJECTS)
 	$(CC) -c $(CFLAGS) $(1) $(SOC_DIRECTORY)/software/libcompiler_rt/mulsi3.c -o mulsi3.o
-	$(AR) crs libcompiler_rt.a $(OBJECTS)
+	$(AR) crs libcompiler_rt.a $(OBJECTS) mulsi3.o
```
Comment (mithro) [inline litex/soc/software/libcompiler_rt/Makefile]: Why is this added?

File: litex/soc/cores/cpu/lm32/verilog/config_lite/lm32_config.v (hunk 1 above)
Comment (cr1901) [reply litex/soc/cores/cpu/lm32/verilog/config_lite/lm32_config.v]: Every maintained synthesizer available today is smart enough to infer block RAM.

But to make you feel better I've explicitly checked `yosy` in the past, and yes it will infer BRAM :).

File: litex/soc/software/libcompiler_rt/Makefile (hunk 5 above)
Comment (cr1901) [reply litex/soc/software/libcompiler_rt/Makefile]: >mulsi3.c was compiled but never linked into libcompiler_rt.a, so I fixed that as well. The minimal variant will choke when linking the BIOS without this.

Meaning I guess no one ever actually checked that cores w/o multipliers worked in `litex`.

File: litex/soc/cores/cpu/lm32/verilog/config_lite/lm32_config.v [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+// available.
+`define CFG_MC_BARREL_SHIFT_ENABLED
+//`define CFG_PL_BARREL_SHIFT_ENABLED
+
+// Multiplier
+// The multiplier is available either in a multi-cycle version or
+// in a pipelined one. The multi-cycle multiplier stalls the pipe
+// for 32 cycles. If both options are disabled, multiply operations
+// are not supported.
+`define CFG_MC_MULTIPLY_ENABLED
```
[the remark this answers is not in the document]
Comment (cr1901) [reply litex/soc/cores/cpu/lm32/verilog/config_lite/lm32_config.v]: Minimal:

```
21.27. Printing statistics.

=== top ===

   Number of wires:               3404
   Number of wire bits:           8588
   Number of public wires:         500
   Number of public wire bits:    4943
   Number of memories:               0
   Number of memory bits:            0
   Number of processes:              0
   Number of cells:               5166
     SB_CARRY                      334
     SB_DFF                         38
     SB_DFFE                       304
     SB_DFFESR                    1051
     SB_DFFESS                      17
     SB_DFFS                         2
     SB_DFFSR                      345
     SB_GB                           1
     SB_LUT4                      3048
     SB_PLL40_CORE                   1
     SB_RAM40_4K                    25
```

Lite:

```
21.27. Printing statistics.

=== top ===

   Number of wires:               3985
   Number of wire bits:           9802
   Number of public wires:         556
   Number of public wire bits:    5482
   Number of memories:               0
   Number of memory bits:            0
   Number of processes:              0
   Number of cells:               5897
     SB_CARRY                      402
     SB_DFF                        108
     SB_DFFE                       334
     SB_DFFESR                    1119
     SB_DFFESS                      25
     SB_DFFS                         2
     SB_DFFSR                      350
     SB_GB                           1
     SB_LUT4                      3524
     SB_PLL40_CORE                   1
     SB_RAM40_4K                    31
```

Both using SPIflash boot.

File: litex/soc/cores/cpu/lm32/verilog/config_lite/lm32_config.v [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+// Instruction cache
+`define CFG_ICACHE_ENABLED
+`define CFG_ICACHE_ASSOCIATIVITY   1
+`define CFG_ICACHE_SETS            128
+`define CFG_ICACHE_BYTES_PER_LINE  16
+`define CFG_ICACHE_BASE_ADDRESS    32'h00000000
+`define CFG_ICACHE_LIMIT           32'h7fffffff
+
+// Data cache
+//`define CFG_DCACHE_ENABLED
```
Comment (mithro) [inline litex/soc/cores/cpu/lm32/verilog/config_lite/lm32_config.v]: This being commented out means the DCACHE is disabled despite the flags below?

Comment (cr1901) [reply litex/soc/cores/cpu/lm32/verilog/config_lite/lm32_config.v]: Correct.

File: litex/soc/software/libcompiler_rt/Makefile (hunk 5 above)
Comment (cr1901) [reply litex/soc/software/libcompiler_rt/Makefile]: Yes, and `libgcc` (or whatever spec `libgcc` follows) mandates it, but `compiler_rt` didn't have it until [February of this year](https://reviews.llvm.org/D42958), and only for RISC-V at that.

**Comment by mithro** [comment]
I'm going to merge, we can tweak more in the future.
